File: typescript/src/scenarios/s12-message-groups.ts
import { AckHandler, IMessageBus, IncomingMessage } from '../abstractions';
import {
  BusScenario,
  ScenarioResult,
  fail,
  nonce,
  pass,
  unsupported,
  waitUntil,
} from './scenario';

const GROUPS = ['a', 'b', 'c'];
const PER_GROUP = 6; // ordered sequence 0..PER_GROUP-1 per group
const TOTAL = GROUPS.length * PER_GROUP;

/**
 * S12 — Ordered delivery / message groups. Messages of several groups are
 * published *interleaved* to ONE topic consumed by two competing consumers
 * (same `subscriberId`) with `partitionByGroup`. The broker must pin each group
 * to a single consumer so that, despite work-sharing, every group's messages
 * arrive in their published order and a group is never split across consumers.
 *
 * Artemis: native message groups (`_AMQ_GROUP_ID`) pin a group to one consumer
 * on the shared multicast queue. RabbitMQ: a consistent-hash exchange routes a
 * `groupId` to a fixed per-consumer queue (needs the bundled plugin) — gated by
 * `supportsMessageGroups`. In-memory: group→consumer affinity on first sight.
 */
export const messageGroups: BusScenario = {
  name: 'S12 message groups',
  description: `${TOTAL} messages across ${GROUPS.length} groups keep per-group order, each pinned to one of 2 consumers.`,
  async run(bus: IMessageBus): Promise<ScenarioResult> {
    const t0 = performance.now();
    if (!bus.busCapabilities.supportsManualAck) {
      return unsupported(this.name, 'no manual ack on this broker', t0);
    }
    if (bus.busCapabilities.supportsMessageGroups === false) {
      return unsupported(this.name, 'no broker-native message grouping', t0);
    }
    const topic = `mbc.s12.${nonce()}`;
    const queueId = `groups-${nonce()}`; // shared queue → competing consumers

    // Record, per consumer, the (group, seq) of each message it handled, in
    // arrival order — enough to assert both per-group FIFO and group→consumer
    // pinning.
    const byConsumer: { group: string; seq: number }[][] = [[], []];

    const make = (i: number): AckHandler => async (m: IncomingMessage) => {
      const [group, seqStr] = m.body.split(':');
      byConsumer[i].push({ group, seq: Number(seqStr) });
      await m.ack();
    };
    const subA = await bus.subscribe(topic, make(0), {
      subscriberId: queueId,
      partitionByGroup: true,
    });
    const subB = await bus.subscribe(topic, make(1), {
      subscriberId: queueId,
      partitionByGroup: true,
    });
    try {
      // Publish interleaved: a0,b0,c0,a1,b1,c1,… so ordering is only preserved
      // if the broker actually pins each group to one consumer.
      for (let seq = 0; seq < PER_GROUP; seq++) {
        for (const g of GROUPS) {
          await bus.publish(topic, `${g}:${seq}`, undefined, { groupId: g });
        }
      }

      const total = () => byConsumer[0].length + byConsumer[1].length;
      const ok = await waitUntil(() => total() >= TOTAL, 10000);
      if (!ok) {
        return fail(this.name, `received ${total()}/${TOTAL}`, t0);
      }

      // Which consumer(s) handled each group, and in what order per consumer.
      const ownersOf = new Map<string, Set<number>>();
      for (let i = 0; i < byConsumer.length; i++) {
        // Per-group sequences as this consumer saw them, in arrival order.
        const perGroupSeqs = new Map<string, number[]>();
        for (const { group, seq } of byConsumer[i]) {
          if (!ownersOf.has(group)) ownersOf.set(group, new Set());
          ownersOf.get(group)!.add(i);
          if (!perGroupSeqs.has(group)) perGroupSeqs.set(group, []);
          perGroupSeqs.get(group)!.push(seq);
        }
        // Within a consumer, each group's sequences must be strictly increasing.
        for (const [group, seqs] of perGroupSeqs) {
          for (let k = 1; k < seqs.length; k++) {
            if (seqs[k] <= seqs[k - 1]) {
              return fail(
                this.name,
                `group ${group} out of order on consumer ${i}: ${seqs.join(',')}`,
                t0,
              );
            }
          }
        }
      }

      // Every group must have been handled by exactly one consumer (pinned).
      for (const g of GROUPS) {
        const owners = ownersOf.get(g);
        if (!owners || owners.size === 0) {
          return fail(this.name, `group ${g} was never delivered`, t0);
        }
        if (owners.size > 1) {
          return fail(this.name, `group ${g} split across consumers`, t0);
        }
        // And the consumer must have seen the full 0..PER_GROUP-1 run.
        const seen = byConsumer[[...owners][0]]
          .filter((m) => m.group === g)
          .map((m) => m.seq);
        if (seen.length !== PER_GROUP) {
          return fail(
            this.name,
            `group ${g} got ${seen.length}/${PER_GROUP} messages`,
            t0,
          );
        }
      }

      const split = byConsumer.map((c) => c.length).join('/');
      return pass(
        this.name,
        `per-group order preserved, each group pinned to one consumer (split ${split})`,
        t0,
      );
    } finally {
      await subA.unsubscribe();
      await subB.unsubscribe();
    }
  },
};

File: dotnet/src/Messaging.Scenarios/S12MessageGroups.cs
using Messaging.Abstractions;
using static Messaging.Scenarios.ScenarioHelpers;

namespace Messaging.Scenarios;

/// <summary>
/// S12 — Ordered delivery / message groups. Messages of several groups are
/// published <i>interleaved</i> to ONE topic consumed by two competing consumers
/// (same <c>SubscriberId</c>) with <c>PartitionByGroup</c>. The broker must pin
/// each group to a single consumer so that, despite work-sharing, every group's
/// messages arrive in their published order and a group is never split.
///
/// Artemis: native message groups (AMQP <c>group-id</c>) pin a group to one
/// consumer on the shared multicast queue. RabbitMQ: a consistent-hash exchange
/// routes a groupId to a fixed per-consumer queue (bundled plugin) — gated by
/// <see cref="BusCapabilities.SupportsMessageGroups"/>. In-memory: group→consumer
/// affinity on first sight.
/// </summary>
public sealed class MessageGroups : IBusScenario
{
    private static readonly string[] Groups = { "a", "b", "c" };
    private const int PerGroup = 6; // ordered sequence 0..PerGroup-1 per group
    private static readonly int Total = Groups.Length * PerGroup;

    public string Name => "S12 message groups";
    public string Description =>
        $"{Total} messages across {Groups.Length} groups keep per-group order, each pinned to one of 2 consumers.";

    public async Task<ScenarioResult> RunAsync(IMessageBus bus)
    {
        var t0 = StartClock();
        if (!bus.BusCapabilities.SupportsManualAck)
            return Unsupported(Name, "no manual ack on this broker", t0);
        if (!bus.BusCapabilities.SupportsMessageGroups)
            return Unsupported(Name, "no broker-native message grouping", t0);

        var topic = $"mbc.s12.{Nonce()}";
        var queueId = $"groups-{Nonce()}"; // shared queue → competing consumers
        var gate = new object();
        // Per consumer, the (group, seq) of each message it handled, in arrival order.
        var byConsumer = new[] { new List<(string Group, int Seq)>(), new List<(string Group, int Seq)>() };

        AckHandler Make(int i) => async m =>
        {
            var parts = m.Body.Split(':');
            lock (gate) byConsumer[i].Add((parts[0], int.Parse(parts[1])));
            await m.AckAsync();
        };

        var subA = await bus.SubscribeAsync(topic, Make(0),
            new SubscribeOptions { SubscriberId = queueId, PartitionByGroup = true });
        var subB = await bus.SubscribeAsync(topic, Make(1),
            new SubscribeOptions { SubscriberId = queueId, PartitionByGroup = true });
        try
        {
            // Publish interleaved: a0,b0,c0,a1,b1,c1,… so ordering is only preserved
            // if the broker actually pins each group to one consumer.
            for (var seq = 0; seq < PerGroup; seq++)
                foreach (var g in Groups)
                    await bus.PublishAsync(topic, $"{g}:{seq}", options: new PublishOptions { GroupId = g });

            int Total2() { lock (gate) return byConsumer[0].Count + byConsumer[1].Count; }
            var ok = await WaitUntilAsync(() => Total2() >= Total, TimeSpan.FromSeconds(10));
            if (!ok)
                return Fail(Name, $"received {Total2()}/{Total}", t0);

            List<(string Group, int Seq)>[] snapshot;
            lock (gate) snapshot = new[] { byConsumer[0].ToList(), byConsumer[1].ToList() };

            // Which consumer(s) handled each group; per-group order per consumer.
            var ownersOf = new Dictionary<string, HashSet<int>>();
            for (var i = 0; i < snapshot.Length; i++)
            {
                var perGroupSeqs = new Dictionary<string, List<int>>();
                foreach (var (group, seq) in snapshot[i])
                {
                    if (!ownersOf.TryGetValue(group, out var owners)) ownersOf[group] = owners = new HashSet<int>();
                    owners.Add(i);
                    if (!perGroupSeqs.TryGetValue(group, out var seqs)) perGroupSeqs[group] = seqs = new List<int>();
                    seqs.Add(seq);
                }
                foreach (var (group, seqs) in perGroupSeqs)
                    for (var k = 1; k < seqs.Count; k++)
                        if (seqs[k] <= seqs[k - 1])
                            return Fail(Name, $"group {group} out of order on consumer {i}: {string.Join(",", seqs)}", t0);
            }

            // Every group handled by exactly one consumer, and it saw the full run.
            foreach (var g in Groups)
            {
                if (!ownersOf.TryGetValue(g, out var owners) || owners.Count == 0)
                    return Fail(Name, $"group {g} was never delivered", t0);
                if (owners.Count > 1)
                    return Fail(Name, $"group {g} split across consumers", t0);
                var owner = owners.First();
                var seen = snapshot[owner].Where(m => m.Group == g).Select(m => m.Seq).ToList();
                if (seen.Count != PerGroup)
                    return Fail(Name, $"group {g} got {seen.Count}/{PerGroup} messages", t0);
            }

            var split = string.Join("/", snapshot.Select(c => c.Count));
            return Pass(Name, $"per-group order preserved, each group pinned to one consumer (split {split})", t0);
        }
        finally
        {
            await subA.DisposeAsync();
            await subB.DisposeAsync();
        }
    }
}
